### DISCUSSION TITLE
Drag&Drop or file input element support in Browsers

### DISCUSSION THREAD

**Discussion by rebior** [donmccurdy/glTF-Transform#526] [Q&A] [answered]
Hello,
when I drag&drop a gltf file together with all the assets into a browser I get a [FileList](https://developer.mozilla.org/en-US/docs/Web/API/FileList). Is it possible to load all of these with WebIO? I had a brief look into the documentation and couldn't find an obvious interface for that. The problem is that the files cannot be fetched from a server. three.js makes that possible by using a loader manager, where you can translate the referenced asset path to some other url. Creating an [object URL](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL) I can easily inject the files that were dropped together with the gltf.

**Comment by donmccurdy** [comment, marked as answer]
The general approach would be to assemble a [JSONDocument](https://gltf-transform.donmccurdy.com/interfaces/core.jsondocument.html) object containing:

- `.json` property (`.gltf` JSON content) 
- `.resources` property (mapping URLs/paths to binary Uint8Array objects) 

Or if you have a `.glb` you can, invoke `io.readBinary(glb)` on the Uint8Array directly.

Those methods could be used to support drag-and-drop (I've implemented [an example here](https://github.com/donmccurdy/glTF-Transform-Render/blob/52431daf421aab23a426eabad8e86dbbcab25dca/examples/dropzone.ts#L45-L65)), or to support any other source the I/O classes don't handle out of the box like websockets, complex authentication, database connections, etc.

**Comment by rebior** [comment]
Thank you for your quick response!
An IO class for FileList would be nice. Maybe I create one and make a PR...

**Reply by donmccurdy** [reply]
Let me know if you have trouble getting that working! I probably won't plan to include FileList input directly in the library. With three I/O classes already taking time to maintain (for Web, Node, and Deno environments), I think more would be too much maintenance vs. spending time to maintain the core functions of the library.
